txjuju/testing/api.py

```diff
@@ -195,20 +197,48 @@ def _formatAnnotationInfo(self, info, verb):
 
     def _formatApplicationInfo(self, info, verb):
         if self.version.startswith("2."):
-            return ["application", verb, {
+            formatted = {
                 "name": info.name,
-                "charm-url": info.charmURL}]
+                "exposed": info.exposed,
+                }
+            if info.charmURL is not None:
+                formatted["charm-url"] = info.charmURL
```

Comment:
Why include the key conditionally instead of unconditionally (as in the original version)?

It seems that not including the key depending on whether is not or None makes life slightly harder for call sites, because they will always need to check for the presence of the key and/or use .get().

In general, the more safe assumptions consumer code can make about a data structure the better it is, I think.